### PR TITLE
Phase diagram

### PR DESCRIPTION
**Pull request by lhy11009**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

I add a cookbook to illustrate how to visualize a phase diagram implemented in a material model from model results.

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

**Comment by lhy11009** [comment]
@jdannberg Thanks for your previous review.

To your questions
_Was there something that wasn't working with these models as the comments indicates, or is this fixed now?_ 
Yes, this is fixed right now so that I removed the comments in the prm file.
 _"Didn't you want to read in the table directly in the visco-plastic model?"_
Right now I implemented this in my version of ViscoPlastic model. 
If this is helpful for the ViscoPlastic model in the mater branch, I can take some time to do that in a separate PR. It should be straightforward.
_I think it would be nice to have one or two sentences about how the phase diagrams compare, but I will leave this up to you._
I have added two lines of comparison between the two examples.
_Can you clarify what you mean by this comment? Do you mean "the thermal expansivity needs to be zero"? Since it is already set to 0.0_
I change that to "set thermal expansivity to 0.0"

File: doc/modules/changes/20210716_lhy11009 [hunk 1]
```diff
@@ -0,0 +1,4 @@
+New: There is now a cookbook that visualizes phase diagram from results of model runs.
```
Comment (tjhei) [inline doc/modules/changes/20210716_lhy11009:1]:
```suggestion
New: There is now a cookbook that visualizes the phase diagram from results of a model run.
```

File: doc/modules/changes/20210716_lhy11009 [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+This includes examples from ViscoPlastic and steinberger material model
```
Comment (tjhei) [inline doc/modules/changes/20210716_lhy11009:2]:
```suggestion
This includes examples from the Visco-Plastic and Steinberger material model.
```

File: doc/modules/changes/20210716_lhy11009 [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+<br>
+(Haoyuan Li, 2021/07/16)
```
Comment (tjhei) [inline doc/modules/changes/20210716_lhy11009:4]: Maybe add Magali here?

File: doc/manual/manual.tex [hunk 4]
```diff
@@ -5907,8 +5907,16 @@ \subsubsection{Using lazy expression syntax for if-else-statements in function e
 Finally, in order to just test whether the initial temperature structure has been properly defined, it is helpful to run the model for a single time-step and without actually waiting for the solvers to converge. In order to do this, the \texttt{End time} can be set to zero. If the model is very large (lots of refinement) or there are large viscosity jumps that take longer to converge in the Stokes solver, it can also be useful to set the solver tolerance parameters to large values as noted below. However, remember that in that case the solution will not be converged -- it is only useful for checking that the initial condition is set correctly.
 \lstinputlisting[language=prmfile]{cookbooks/muparser-temperature-example/doc/check-initial-condition.part.prm.out}
 
+<<<<<<< HEAD
 \input{cookbooks/christensen_yuen_phase_function/doc/christensen_yuen_phase_function.tex}
 
+% cookbooks/phase_diagram/
+||||||| parent of a7445755b... Update doc/manual/manual.tex
```
Comment (tjhei) [inline doc/manual/manual.tex:5914]: You need to resolve the conflicts...

**Review by tjhei — CHANGES_REQUESTED** [review]

**Comment by lhy11009** [comment]
Hi @tjhei, thanks for your review, I have accepted your requested changes

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+Pressure, on the other hand, increases from the top of the box to the bottom with a constant gradient.
+This is assured by assigning a constant density with zero expansivity.
+It differs from a classic phase diagram only in the direction of pressure increase.
+
+\par Two compositions of pyrolite and harzburgite are included in the model. % composition
+With each run, only one composition is assigned to the whole domain in order to visualize the diagrams of these two separately.
+
+\par The material model is then set up to mimic mantle phase transitions at 410, 520, and 660 km.% setup of mantle transitions
+Details of phase transitions are taken from \cite{billen2018decoupling}.
+A trick is needed to make this work: I assign the values of the density to the field of thermal capacity in the input file. % trick
```
Comment (tjhei) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:25]:
```suggestion
A trick is needed to make this work: we assign the values of the density to the field of thermal capacity in the input file. % trick
```
Do you mean thermal expansivity?

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+To do this, I used the phase inputs implemented in the ViscoPlastic file.
```
Comment (tjhei) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:26]: What do you mean by file?

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+This serves the goal of visualizing values of reference densities of phases with the 'real density' assigned with a constant value in the model.
```
Comment (tjhei) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:27]: Are these quotes correct for latex? I would think not.

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+Inputs for this material model is listed here:
```
Comment (tjhei) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:28]:
```suggestion
Inputs for this material model are listed here:
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+
+\paragraph{Results}
+
+\par Visualization of the model results yields a phase diagram of a pyrolitic mantle (Figure \ref{fig:phase_diagram_ph_density}). % phase diagram
+The field shown here has the reference densities of the pyrolite phases, though settings of phase transitions are over-simplified.
+One may notice that 3 transitions (i.e. one in the olivine system, two in the spinel system) are included for the 660 interfaces, % 3 transitons, higher T
+and they need to be modified at a higher temperature.
+In spite of the complexities of mantle phases, the focus of this first example is to simply illustrate this approach of visualizing it.
+Beyond showing the diagram, I have also used the 'lineout' feature in VISIT to export the data along two vertical lines at T = 1173 and T = 1673 (Figure \ref{fig:phase_diagram_ph_profile}). % linear profile
```
Comment (tjhei) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:41]: Here and everywhere: use we instead of I. Also, check the quotes here as well.

**Review by tjhei — COMMENTED** [review]
I have some minor stylistic comments. I would use "we" instead of "I" throughout, for example.

I have one more fundamental question. You are setting up a fake box geometry with fixed T and p to generate a phase diagram. Wouldn't it be easier for the user id you write a postprocessor that generates this diagram for an arbitrary model run by evaluating the material model on a p, T grid? Let me know if you want me to explain in more detail...

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex (hunk 5 above)
Comment (lhy11009) [reply doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex]: It's heat capacity, I have changed that

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex (hunk 6 above)
Comment (lhy11009) [reply doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:26]: I have changed to plugin, also ViscoPlastic should be visco plastic

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex (hunk 9 above)
Comment (lhy11009) [reply doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:41]: In this case, the quotes seem unnecessary, I deleted them

File: doc/manual/cookbooks/phase_diagram/doc/harzburgite.prm [hunk 10]
```diff
@@ -0,0 +1,8 @@
+# initial composition model
+subsection Initial composition model
+    set List of model names = function
+    subsection Function
+        set Coordinate system = cartesian
+        set Function expression = 1.0  # change this to one
+    end
+end
```
[the remark this answers is not in the document]
Comment (lhy11009) [reply doc/manual/cookbooks/phase_diagram/doc/harzburgite.prm:8]: I have fixed that.

File: doc/manual/cookbooks/phase_diagram/doc/material_model.prm [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+        set Maximum viscosity = 1e24
+        set Phase transition depths = background:410e3|520e3|560e3|670e3|670e3|670e3|670e3, spharz: 410e3|520e3|560e3|670e3|670e3|670e3|670e3
+        set Phase transition widths = background:5e3|5e3|5e3|5e3|5e3|5e3|5e3, spharz: 5e3|5e3|5e3|5e3|5e3|5e3|5e3
+        set Phase transition temperatures = background:1662.0|1662.0|1662.0|1662.0|1662.0|1662.0|1662.0, spharz: 1662.0|1662.0|1662.0|1662.0|1662.0|1662.0|1662.0
+        set Phase transition Clapeyron slopes = background:4e6|4.1e6|4e6|-2e6|4e6|-3.1e6|1.3e6, spharz: 4e6|4.1e6|4e6|-2e6|4e6|-3.1e6|1.3e6
+        set Densities = 3300.0
+        set Heat capacities = background: 3300.0|3394.4|3442.1|3453.2|3617.6|3691.5|3774.7|3929.1, spharz: 3235.0|3372.3|3441.7|3441.7|3680.8|3717.8|3759.4|3836.6
+        set Thermal expansivities = 0.0
+    end
+end
```
[the remark this answers is not in the document]
Comment (lhy11009) [reply doc/manual/cookbooks/phase_diagram/doc/material_model.prm:14]: same

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 12]
```diff
@@ -0,0 +1,83 @@
+\subsubsection{Phase diagram}
+
+\label{sec:phase-diagram}
+For many models it is useful to see how the material parameters vary as a function of pressure and temperature. In this cookbook, a plot of material properties (e.g., density or viscosity) as a function of temperature (x axis) and pressure (y axis) is referred to as a phase diagram. 
+
+In this phase diagram cookbook, phase diagrams are generated from model runs and visualized with VISIT, but Paraview could also be used to view the same output. % intro
+This is a simple way of using the visualization output to check the correct implementation of phase transitions in a material model. 
+The idea lies in using a model with an initial temperature that increases linearly along the x-axis and a pressure that changes linearly along the y-axis in a box geometry, so that the two axes, x and y, can be used as axes for pressure and temperature in the graphical output. 
+Here, I visualize a diagram of the phase transitions implemented in the visco plastic material model, as well as a lookup table in the Steinberger material model.
```
[the remark this answers is not in the document]
Comment (lhy11009) [reply doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex]: I have changed all the 'I' to 'we'

**Comment by lhy11009** [comment]
> I have some minor stylistic comments. I would use "we" instead of "I" throughout, for example.
> 
> I have one more fundamental question. You are setting up a fake box geometry with fixed T and p to generate a phase diagram. Wouldn't it be easier for the user id you write a postprocessor that generates this diagram for an arbitrary model run by evaluating the material model on a p, T grid? Let me know if you want me to explain in more detail...

Hi Timo, this idea of a new post-processor is interesting to me. It would be helpful if you're willing to explain it to me in more detail.
I think in terms of the logistics, we need to
 

- generate a new grid in P and T
- call the eos somewhere. I guess this has to be in a separate material model.
- feed that to the post-processor.

Anyway, I'll try to make tomorrow's biannual meeting, maybe explain it to me when we have a chance?

**Comment by lhy11009** [comment]
@tjhei Did you see my request to review? Sorry if I am checking when you are too busy. But it really just needs a final view. Your new idea is also interesting, let's chat about that when having a chance.

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+In this phase diagram cookbook, phase diagrams are generated from model runs and visualized with VISIT, but Paraview could also be used to view the same output. % intro
+This is a simple way of using the visualization output to check the correct implementation of phase transitions in a material model. 
+The idea lies in using a model with an initial temperature that increases linearly along the x-axis and a pressure that changes linearly along the y-axis in a box geometry, so that the two axes, x and y, can be used as axes for pressure and temperature in the graphical output. 
+Here, we visualize a diagram of the phase transitions implemented in the visco plastic material model, as well as a lookup table in the Steinberger material model. 
+
+\paragraph{The input file.}
+You can find the input file to run this cookbook example in \url{cookbooks/phase_diagram.prm}. For this first case, phase transitions are prescribed manually in terms of their depth, Clapeyron slope, and other key parameters. 
+
+\par The model domain is  800 km  by 800 km box. % geometry
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:14]: Here and below, drop the `\par` commands -- let latex figure out the layout of text.
```suggestion
The model domain is  800 km  by 800 km box. % geometry
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+Initial temperature increases from 273 K on the left side to 2273 K on the right side. % initial conditions
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:15]:
```suggestion
Initial temperature increases from \SI{273}{K} on the left side to \SI{2273}{K} on the right side. % initial conditions
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+Pressure, on the other hand, increases from the top of the box to the bottom with a constant gradient.
+This is assured by assigning a constant density with zero expansivity.
+It differs from a classic phase diagram only in the direction of pressure increase.
+
+\par Two compositions of pyrolite and harzburgite are included in the model. % composition
+With each run, only one composition is assigned to the whole domain in order to visualize the diagrams of these two separately.
+
+\par The material model is then set up to mimic mantle phase transitions at 410, 520, and 660 km.% setup of mantle transitions
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:23]:
```suggestion
\par The material model is then set up to mimic mantle phase transitions at 410, 520, and \SI{660}{km}.% setup of mantle transitions
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+Details of phase transitions are taken from \cite{billen2018decoupling}.
+A trick is needed to make this work: we assign the values of the density to the field of heat capacity in the input file. % trick
+To do this, We used the phase inputs implemented in the visco plastic plug-in.
+This serves the goal of visualizing values of reference densities of phases with the real density assigned with a constant value in the model.
+Inputs for this material model are listed here:
+\lstinputlisting[language=prmfile]{cookbooks/phase_diagram/doc/material_model.prm.out}
+Compared to the first example, pyrolitic phases are illustrated on a much finer scale.  % comparison 
+Meanwhile, we can still tell the phase transitions at 410, 520, and 660 km depth, respectively, marked by linear boundaries analogous to a constant Clapeyron slope.
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:31]:
```suggestion
Meanwhile, we can still tell the phase transitions at 410, 520, and \SI{660}{km} depth, respectively, marked by linear boundaries analogous to a constant Clapeyron slope.
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+
+\paragraph{Results}
+
+\par Visualization of the model results yields a phase diagram of a pyrolitic mantle (Figure \ref{fig:phase_diagram_ph_density}). % phase diagram
+The field shown here has the reference densities of the pyrolite phases, though settings of phase transitions are over-simplified.
+One may notice that 3 transitions (i.e. one in the olivine system, two in the spinel system) are included for the 660 interfaces, % 3 transitons, higher T
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:38]:
```suggestion
One may notice that three transitions (i.e., one in the olivine system, two in the spinel system) are included for the 660 interfaces, % 3 transitons, higher T
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+and they need to be modified at a higher temperature.
+In spite of the complexities of mantle phases, the focus of this first example is to simply illustrate this approach of visualizing it.
+Beyond showing the diagram, We have also used the lineout feature in VISIT to export the data along two vertical lines at T = 1173 and T = 1673 (Figure \ref{fig:phase_diagram_ph_profile}). % linear profile
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:41]:
```suggestion
Beyond showing the diagram, We have also used the ``lineout'' feature in VisIt to export the data along two vertical lines at $T = \SI{1173}{K}$ and $T = \SI{1673}{K} (Figure \ref{fig:phase_diagram_ph_profile}). % linear profile
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+The T = 1173 illustrates the buoyancy forces felt by a descending cold slab within the mantle transition zone.
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:42]:
```suggestion
The figure for $T = \SI{1173}{K}$ illustrates the buoyancy forces felt by a descending cold slab within the mantle transition zone.
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+\par Next, we shift to harzburgite by changing the values of the initial compositional field from 0 to 1.  % harzburgite
+For this to occur, following changes are needed:
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:45]:
```suggestion
For this to occur, the following changes are needed:
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+\lstinputlisting[language=prmfile]{cookbooks/phase_diagram/doc/harzburgite.prm.out}
+With this change, we can also visualize the phase diagram of harzburgite in Figure \ref{fig:phase_diagram_ph_density}.
+
+\par Moreover, We tested the a pyrolitic lookup table used in the Steinberg material model (Figure \ref{fig:phase_diagram_steinberg_density}).% steinberg model
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:49]:
```suggestion
\par Moreover, We tested the pyrolitic lookup table used in the Steinberg material model (Figure \ref{fig:phase_diagram_steinberg_density}).% steinberg model
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+
+%%%%% figures %%%%%%
+%%%%%%%%% figures1: pyrolite %%%%%%
+\begin{figure}
+\phantom.
+\hfill
+\includegraphics[width=0.9\textwidth]{cookbooks/phase_diagram/doc/pyrolite_harzburgite.png}
+\hfill
+\phantom.
+\caption{\it The field of heat capacity showing values of reference densities for Pyrolitic and Harzburgitic phases.}
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:66]: I'm not a geologist, but I think that everywhere else we use lower-case for rock types, right?
```suggestion
\caption{\it The field of heat capacity showing values of reference densities for pyrolitic and harzburgitic phases.}
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+\label{fig:phase_diagram_ph_density}
+\end{figure}
+
+%%%%%%%%%% figures2: pyrolite profiles %%%%%%
+\begin{figure}
+\centering
+\includegraphics[width=0.4\textwidth]{cookbooks/phase_diagram/doc/pyrolite_linear.png}
+\caption{\it Profiles of pyrolitic density at T=1173K(red) and 1673K(blue).}
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:74]:
```suggestion
\caption{\it Profiles of pyrolitic density at $T=\SI{1173}{K}$ (red) and $\SI{1673}{K}$ (blue).}
```

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+\label{fig:phase_diagram_ph_profile}
+\end{figure}
+%%%%%%%%% figures3: lookup table %%%%%%
+\begin{figure}
+\centering
+\includegraphics[width=0.4\textwidth]{cookbooks/phase_diagram/doc/steinberg.png}
+\caption{\it Density from lookup table of pyrolite from \cite{stixrude2011thermodynamics}} % cite
```
Comment (bangerth) [inline doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:81]:
```suggestion
\caption{\it Density from lookup table of pyrolite from \cite{stixrude2011thermodynamics}.} % cite
```

**Review by bangerth — COMMENTED** [review]
That's a really nice idea! Almost all of my comments are just small text issues.

**Comment by bangerth** [comment]
I think @tjhei 's idea has value, but I like the current form for its clever use of existing infrastructure. It's an easy way to achieve what you want without having to go to greater lengths. Writing a postprocessor may come in a separate step.

If you address my comments, please also squash the commits. I think you're very close to this being ready.

**Review by tjhei — APPROVED** [review]
Please address Wolfgang's comments. I am okay to merge afterwards.
(Sorry for accidentally clicking "close')

**Comment by tjhei** [comment]
When you are done, you will need to rebase to the most recent master version as well.

File: cookbooks/phase_diagram.prm [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+set Dimension                              = 2
+set CFL number                             = 1.0
+set End time                               = 0
+set Adiabatic surface temperature          = 1673.0
+set Output directory                       = output
+set Use years in output instead of seconds = true
+
+# Do not solve the Stokes system.
+subsection Solver parameters
+  subsection Stokes solver parameters
```
[the remark this answers is not in the document]
Comment (lhy11009) [reply cookbooks/phase_diagram.prm]: I shifted to 'no Advection, no Stokes' and deleted this section.

File: cookbooks/phase_diagram.prm [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+#    set Use lateral average temperature for viscosity = true
+#    set Number lateral average bands                  = 10
+#
+#    set Reference viscosity                           = 1e21
+#    set Bilinear interpolation                        = true
+#    set Latent heat                                   = false
+#  end
+#end
+
+# comment the following 3 sections for steinberg model and pyrolitic lookup table
```
[the remark this answers is not in the document]
Comment (lhy11009) [reply cookbooks/phase_diagram.prm]: I have changed them all to start with capital letters.

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 27]
```diff
@@ -0,0 +1,83 @@
+\subsubsection{Phase diagram}
+
+\label{sec:phase-diagram}
+For many models it is useful to see how the material parameters vary as a function of pressure and temperature. In this cookbook, a plot of material properties (e.g., density or viscosity) as a function of temperature (x axis) and pressure (y axis) is referred to as a phase diagram.
```
[the remark this answers is not in the document]
Comment (lhy11009) [reply doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:4]: I have added me and Magali as authors of this cookbook and me as an author of the mannual.

**Comment by lhy11009** [comment]
@bangerth Thanks for your review. I have resolved all conversations. And @tjhei, it's okay, I haven't even noticed it was closed.

**Comment by bangerth** [comment]
OK, last step: squash the commits into one. Do you know how to do that?

**Comment by lhy11009** [comment]
@bangerth I think I have done that just now.

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 28]
```diff
@@ -0,0 +1,83 @@
+\subsubsection{Phase diagram}
```
[the remark this answers is not in the document]
Comment (tjhei) [reply doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:1]: What about this change?

File: doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+
+\label{sec:phase-diagram}
+For many models it is useful to see how the material parameters vary as a function of pressure and temperature. In this cookbook, a plot of material properties (e.g., density or viscosity) as a function of temperature (x axis) and pressure (y axis) is referred to as a phase diagram. 
+
+In this phase diagram cookbook, phase diagrams are generated from model runs and visualized with VISIT, but Paraview could also be used to view the same output. % intro
```
[the remark this answers is not in the document]
Comment (tjhei) [reply doc/manual/cookbooks/phase_diagram/doc/phase_diagram.tex:6]: It looks like you are missing this and many other changes for m Wolfgang. Maybe you forgot to pull from GitHub before squashing your commits? If you push you will overwrite your changes.

**Comment by lhy11009** [comment]
@tjhei I saw you commented 5 hrs ago by an email. That's odd since I rechecked the branch, by which I first pull from my remote branch and found all of Wolfgang's comments are addressed there. How do I check this further? Can you point me to where you are looking at. By the way, your comment didn't show up on this github page, which is also confusing.

**Comment by tjhei** [comment]
If you go to the PR on GitHub, click on files, and scroll to the .Tex file, to should see Wolfgang's andy comments. Take the title of the section for example.
Or is GitHub somehow broken?

**Comment by bangerth** [comment]
I think @tjhei is right. For example, this comment: https://github.com/geodynamics/aspect/pull/4257/files#diff-60ab888dc4bb054c4a202327f263c788ea40e1b83e6d747d378f659d17b47e40

**Comment by lhy11009** [comment]
@tjhei Yes, you are right. I just rechecked and I can see the unresolved ones in the 'Files changed" tab. Sorry for leaving these unresolved, maybe I have accidentally clicked 'resolve'. Anyway, I have committed those changes and managed them in a single commit.

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Thanks, @lhy11009 . The automated tests now complain about this issue:
```
make: *** [manual.pdf] Error 1

    Error: There were errors compiling manual.tex.
           Check manual/manual.log for details

Makefile:101: recipe for target 'manual.pdf' failed
```
Can you try and generate `manual.pdf` locally on your machine?

**Comment by lhy11009** [comment]
@It complains about the issue that one citation I put in the cookbook cannot be found:
`LaTeX Warning: Citation `billen2018decoupling' on page 135 undefined on input l
ine 24.
`
But I have put that into the manual.bib file:
```
@article{billen2018decoupling,
  title={Decoupling of plate-asthenosphere motion caused by non-linear viscosity during slab folding in the transition zone},
  author={Billen, Magali I and Arredondo, Katrina M},
  journal={Physics of the Earth and Planetary Interiors},
  volume={281},
  pages={17--30},
  year={2018},
  publisher={Elsevier}
}
```
Do you know how to debug this issue?

**Comment by tjhei** [comment]
> I am not sure how to debug this issue

did you use ``make manual.pdf`` in the doc folder to compile?

**Comment by lhy11009** [comment]
Yes, I did use that. I somehow make it work after failed twice.

**Comment by lhy11009** [comment]
I think I am going to debug this by push commits and testing include only a portion of the changes I made in the tex file, this may take me hours and a few days in my current schedule, I'll let you know when I figure this out

**Comment by tjhei** [comment]
No, this is not the way to do it. After you run "make manual pdf", go into the manual/ directory and run "pdflatex manual.tex". It should stop at the error.
If you need help, let us know.

**Comment by lhy11009** [comment]
@tjhei Thank you, I'll remember that. There is an untracked file in my directory. I think that's why I can compile but here it doesn't work. Anyway, I have fixed that and retest it.

**Review by tjhei — APPROVED** [review]

**Pull request merged**